apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,327 @@
+through all the data points and stacks the results. Importantly, it performs this
+operation efficiently by pushing the `for` loop into the PyTorch operations,
+effectively vectorizing the computation.
+
+Consider the following example:
+```python
+import torch
+
+# Written to handle only single sample.
+def my_simple_model(feature_vec, weight):
+    return torch.dot(feature_vec, weight).relu()
+
+batch_size = 4
+batched_inputs = torch.randn(batch_size, 3)
+weight = torch.randn(3)
+
+# For Loop version
+expected = []
+for input in batched_inputs:
+    expected.append(my_simple_model(input, weight))
+expected = torch.stack(expected)
+
+# Vmap
+# `in_dims` specifies the dimension that should be mapped over.
+# In this case, we map only over 0-dim of `batched_inputs`.
+actual = torch.vmap(my_simple_model, in_dims=(0, None))(batched_inputs, weight)
+
+# Verify that the results match.
+torch.testing.assert_close(expected, actual)
+```
+
+To support `vmap` for PyTorch operators, we need to specify the batching rule
+i.e. how to map the given function over a batched input.
+A batching rule is essentially a function which takes one or multiple batched
+inputs and computes the batched operation. In the above example to support
+`vmap` for `my_simple_model`, we need to know the batching rule for `torch.dot`
+and `torch.relu` to be able to vectorize our model.
+PyTorch has a lot of operators and we need to have coverage for all of them
+to support `vmap`. That being said, there is a `for`-loop
+fallback in case an operator is not supported so as not to crash the code.
+
+PyTorch operators can be roughly categorized as primitive (internally
+`CompositeExplicitAutograd`) vs composite (internally `CompositeImplicitAutograd`).
+Primitive operators are the ones for which we specify the batching and
+gradient rules. Composite operators are implemented using these the primitive operators
+and other simpler composite operators. If we implement batching rules for every
+primitive operator, we automatically get the batching rules for composite operators.
+
+There are two ways to add batching support for an operator:
+* Manually write the batching rule. See for example the [batching rule for torch.dot](https://github.com/pytorch/pytorch/blob/b30ee35a6f141d3247a24fd09f96ea50a7e2b3c7/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp#L25-L34)
+* Decompose operators using other operators for which we already have a
+batching rule. See for example the [batching rule for torch.vdot](https://github.com/pytorch/pytorch/blob/b30ee35a6f141d3247a24fd09f96ea50a7e2b3c7/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp#L35-L37)
+
+## Composite Compliance
+
+As mentioned earlier, we obtain batching rules effortlessly for composite operators
+but this holds true only under certain constraints. These constraints include requirements
+like refraining from accessing the tensor's data pointer and avoiding the use of
```

Comment:
"these constraints include requirements like" seems redundant, it could just be "These constraints include refraining from ..."